### PR TITLE
Type check variables between modules

### PR DESCRIPTION
**Pull request by jen20**
This patch makes the tests pass - we expect an error if a string value is assigned to a variable which should be a map. This will need extending for lists/maps etc but is still relevant for the next point release, hence targeting master for now.

### CODE REVIEW

File: terraform/eval_variable.go [hunk 1]
```diff
@@ -2,12 +2,58 @@ package terraform
...
 	"fmt"
+	"strings"
 
 	"github.com/hashicorp/errwrap"
 	"github.com/hashicorp/terraform/config"
+	"github.com/hashicorp/terraform/config/module"
 	"github.com/mitchellh/mapstructure"
 )
 
+type EvalTypeCheckVariable struct {
```
Comment (phinze) [inline terraform/eval_variable.go]: Would you mind doing a prose comment block summarizing the intended use case and context of the EvalNode? I've found these to be helpful when navigating unfamiliar EvalTrees, and the name is often not enough to fully get what the node is supposed to do.

File: terraform/eval_variable.go [hunk 2]
```diff
@@ -2,12 +2,58 @@ package terraform
...
+	Variables  map[string]string
+	ModulePath []string
+	ModuleTree *module.Tree
+	DepPrefix  string
```
Comment (phinze) [inline terraform/eval_variable.go]: I believe `DepPrefix` is vestigal?

File: terraform/eval_variable.go [hunk 3]
```diff
@@ -2,12 +2,58 @@ package terraform
...
+	targetConfig := currentTree.Config()
+
+	prototypes := make(map[string]config.VariableType)
+	for _, variable := range targetConfig.Variables {
+		prototypes[variable.Name] = variable.Type()
+	}
+
+	for name, declaredType := range prototypes {
+		// This is only necessary when we _actually_ check
+		// proposedValue := n.Variables[name]
```
Comment (phinze) [inline terraform/eval_variable.go]: Didn't end up using `proposedValue` it looks like.

File: terraform/eval_variable.go (hunk 2 above)
Comment (jen20) [inline terraform/eval_variable.go]: Ah yes, correct - I missed it from this one.

File: terraform/eval_variable.go (hunk 1 above)
Comment (jen20) [inline terraform/eval_variable.go]: Yes that seems like a worthwhile effort - I think several others can be documented based on the research that went into this PR too!

File: terraform/eval_variable.go [hunk 4]
```diff
@@ -2,12 +2,75 @@ package terraform
...
+// In order to achieve this three things are required:
+//     - a map of the proposed variable values
+//     - the configuration tree of the module in which the variable is
+//       declared
+//     - the path to the module (so we know which part of the tree to
+//       compare the values against).
+//
+// Currently since the type system is simple, we currently do not make
+// use of the values since it is only valid to pass string values. The
+// structure is in place for extension of the type system, however.
```
Comment (phinze) [inline terraform/eval_variable.go]: <3

**Comment by phinze** [comment]
LGTM!

**Pull request merged**